### PR TITLE
Implement freeze handler

### PR DESCRIPTION
**Pull request by alban**
# Implement freeze handler

This is the "freeze" cgroup to freeze the container and allow manual inspection of the container while all processes are suspended.

## How to use

Use a Seccomp Policy with "freeze-container":
```
  listenerPath: "/run/seccomp-agent.socket"
  listenerMetadata: "DEFAULT_ACTION=freeze-container"
```

Try a forbidden syscall:
```
$ kubectl exec -ti mypod -- /bin/sh
/ # unshare -u
```

The container is frozen.

## Testing done

See above.

### CODE REVIEW

File: pkg/handlers/default.go [hunk 1]
```diff
@@ -49,3 +56,68 @@ func KillContainer(pid int) registry.HandlerFunc {
 		return registry.HandlerResultErrno(unix.EPERM)
 	}
 }
+
+func FreezeContainer(pid int) registry.HandlerFunc {
+	re := regexp.MustCompile(`^(\d*):(freezer)?:(/..*)$`)
```
Comment (rata) [inline pkg/handlers/default.go:61]: Why outside the function we return in the next line?

File: pkg/handlers/default.go [hunk 2]
```diff
@@ -49,3 +56,68 @@ func KillContainer(pid int) registry.HandlerFunc {
...
+	return func(fd libseccomp.ScmpFd, req *libseccomp.ScmpNotifReq) (result registry.HandlerResult) {
+		cgroupPath := ""
+		if cgroupFile, err := os.Open(filepath.Join("/proc", fmt.Sprintf("%d", pid), "cgroup")); err == nil {
+			defer cgroupFile.Close()
```
Comment (rata) [inline pkg/handlers/default.go:65]: just curious, are you sure this works? You are defining cgroupFile inside the if, so this var only exists in this scope. Will it be possible to call `cgroupFile.Close()` when the function ends outside of this if? I guess that works? Or panic?

File: pkg/handlers/default.go [hunk 3]
```diff
@@ -49,3 +56,68 @@ func KillContainer(pid int) registry.HandlerFunc {
...
+			reader := bufio.NewReader(cgroupFile)
+			for {
+				line, err := reader.ReadString('\n')
+				if err != nil {
+					break
+				}
+				line = strings.TrimSuffix(line, "\n")
+
+				r := re.FindStringSubmatch(line)
+				if len(r) < 4 {
```
Comment (rata) [inline pkg/handlers/default.go:75]: Why is this? Can you please elaborate and probably add a comment too?

File: pkg/handlers/default.go [hunk 4]
```diff
@@ -49,3 +56,68 @@ func KillContainer(pid int) registry.HandlerFunc {
...
+				cgroupPath = r[3]
+				break
+			}
+		} else {
+			log.WithFields(log.Fields{
+				"pid": pid,
+				"err": err,
+			}).Error("cannot parse cgroup")
+			return registry.HandlerResultErrno(unix.EPERM)
+		}
```
Comment (rata) [inline pkg/handlers/default.go:88]: I think this is more readable if you change the if to be: if err != nil, then do this. No else, and just proceed with the err == nil case. We reduce the indentation and we just filter the "uninteresting cases" at the beginning and continue with what is interesting here (being able to open the file and freeze the container).

File: pkg/handlers/default.go [hunk 5]
```diff
@@ -49,3 +56,68 @@ func KillContainer(pid int) registry.HandlerFunc {
...
+		if cgroupPath == "" {
+			log.WithFields(log.Fields{
+				"pid": pid,
+			}).Error("cgroup path not found")
+			return registry.HandlerResultErrno(unix.EPERM)
+		}
+
+		cgroup := &configs.Cgroup{
```
Comment (rata) [inline pkg/handlers/default.go:96]: configs seems like our config, but it is from libcontainer. Maybe import as `libctConfig` (common prefix used for libcontainer in runc: https://github.com/opencontainers/runc/commits/main/libcontainer)

File: pkg/handlers/default.go [hunk 6]
```diff
@@ -49,3 +56,68 @@ func KillContainer(pid int) registry.HandlerFunc {
...
+			Path:      cgroupPath,
+			Resources: &configs.Resources{},
+		}
+
+		m, err := manager.New(cgroup)
```
Comment (rata) [inline pkg/handlers/default.go:101]: idem, libctManager?

File: pkg/handlers/default.go [hunk 7]
```diff
@@ -49,3 +56,68 @@ func KillContainer(pid int) registry.HandlerFunc {
...
+		if err != nil {
```
Comment (rata) [inline pkg/handlers/default.go:102]: fold the if and the declaration in one line, please (we need a linter!)

File: pkg/handlers/default.go [hunk 8]
```diff
@@ -49,3 +56,68 @@ func KillContainer(pid int) registry.HandlerFunc {
...
+			log.WithFields(log.Fields{
+				"pid":        pid,
+				"cgroupPath": cgroupPath,
+				"err":        err,
+			}).Error("cannot create new cgroup manager")
+			return registry.HandlerResultErrno(unix.EPERM)
+		}
+
+		err = m.Freeze(configs.Frozen)
+		if err != nil {
```
Comment (rata) [inline pkg/handlers/default.go:112]: idem, 1 line

File: pkg/handlers/default.go [hunk 9]
```diff
@@ -49,3 +56,68 @@ func KillContainer(pid int) registry.HandlerFunc {
...
+			log.WithFields(log.Fields{
+				"pid":        pid,
+				"cgroupPath": cgroupPath,
+				"err":        err,
+			}).Error("cannot freeze cgroup")
+			return registry.HandlerResultErrno(unix.EPERM)
+		}
+
+		return registry.HandlerResultErrno(unix.EPERM)
```
Comment (rata) [inline pkg/handlers/default.go:121]: On success we return EPERM? Why? How are the callers using this return value?

File: pkg/handlers/default.go (hunk 1 above)
Comment (rata) [inline pkg/handlers/default.go:61]: What are we matching with this regex? starts with 0 or more numbers, followed by `:`, then `freezer` might be there or not, then `:` and something that starts with `/` and has at least one char (..*)?

Isn't there something simpler we can do, like find the index of the second instance of `:`, and take the substrig from there until the end? Is `:` a char that can appear in the fist two fields?

File: deploy/seccompagent.yaml [hunk 10]
```diff
@@ -65,6 +68,8 @@ spec:
           mountPath: /host/seccomp
         - name: run
           mountPath: /run
+        - name: cgroup
+          mountPath: /sys/fs/cgroup
```
Comment (rata) [inline deploy/seccompagent.yaml:72]: Just curious, I see the code uses `/proc/pid/cgroup` and not this path at all. Why is this needed? Is this needed for that proc file to show the correct content? Or for the libct cgroup freeze to work?

**Review by rata — CHANGES_REQUESTED** [review]
Some questions and changes, overall it looks good but lot of questions about the details :)

File: pkg/handlers/default.go (hunk 3 above)
Comment (rata) [inline pkg/handlers/default.go:75]: Is it guaranteed that the len(r) will be less than 4 some time? Because, otherwise, we have an infinite loop

File: pkg/handlers/default.go (hunk 1 above)
Comment (alban) [reply pkg/handlers/default.go:61]: Just because there is no need to re-compile the regexp for each syscall. We can compile it one time when the container is configured with its seccomp policy.

File: pkg/handlers/default.go (hunk 2 above)
Comment (alban) [reply pkg/handlers/default.go:65]: The variable defined in the 'if' condition is accessible in the 'if' block, including defer statements if necessary (magic of the Go garbage collector).

Note that direct parameters of a defer function are immediately evaluated, even if the defer function is evaluated later:
https://go.dev/play/p/2jHQzzv9Hk1

In this case, cgroupFile is the receiver of the method Close:
```
func (*File) Close
```
So it receives a pointer to a `File`, but receivers are handled in the same way as arguments:
https://go.dev/play/p/EpH1bXIq54-

File: pkg/handlers/default.go (hunk 9 above)
Comment (alban) [reply pkg/handlers/default.go:121]: This PR never thaw the container, so it stays frozen, so this return value is never actually used. But if the alert is sent to an external system, and the container is thaw by that external system, I guess we still don't want the syscall to be allowed.

File: pkg/handlers/default.go (hunk 1 above)
Comment (alban) [reply pkg/handlers/default.go:61]: I will add a comment:
```
	// Matches lines in /proc/$pid/cgroup for
	// - either the freezer cgroup (starting with '%d:freezer:'), or
	// - the unified hierarchy (starting with '0::')
```
So it should work on systems using cgroup-v1, cgroup-v2 or hybrid cgroups.

I want to match at least one character after `/` to make sure it is not the root cgroup (`/`) because that would freeze all processes on the node.

File: deploy/seccompagent.yaml (hunk 10 above)
Comment (alban) [reply deploy/seccompagent.yaml:72]: Yes, the libct cgroup freeze will freeze the container by using /sys/fs/cgroup/freezer (or equivalent, on cgroup-v2 or hybrid cgroups).

I bind mount /sys/fs/cgroup from the host instead of taking the one from the container because:
- this makes sure it is mounted in RW mode, so libcontainer can write "frozen" to it
- in case the container runtime uses cgroupns, so we have the full cgroupfs view by picking the mount from the host (but in the end, we have to make sure that cgroupns is disabled, otherwise /proc/$pid/cgroup would not have absolute paths...)

File: pkg/handlers/default.go (hunk 3 above)
Comment (alban) [reply pkg/handlers/default.go:75]: `reader.ReadString()` returns `io.EOF` at the end of the file, so it will not be an infinite loop.

File: pkg/handlers/default.go (hunk 1 above)
Comment (alban) [reply pkg/handlers/default.go]: I moved it outside of the function, with comments.

**Comment by alban** [comment]
Thanks for the review! I think I addressed all the comments.

File: pkg/handlers/default.go (hunk 2 above)
Comment (rata) [reply pkg/handlers/default.go]: Cool, thanks for the examples and links! :)

File: pkg/handlers/default.go (hunk 1 above)
Comment (rata) [reply pkg/handlers/default.go]: But IIUC `:` is not a valid char for the first field nor the second field, using cgroups v1, v2 or hybrid. Right?

Why not use just the index of the second `:` and just slice it from there? It is way more readable and maintainable, etc. Or is there some edge cases to handle that makes this not so readable?


Also, the regex has several groups today, and we only use one. I think the `(\d)*` can probably just be `\d*`, without grouping, and maybe the second group can be avoided too. But this only applies IF we are using the regex, I personally don't like them and prefer to avoid them if the alternatives are more readable :)

File: deploy/seccompagent.yaml (hunk 10 above)
Comment (rata) [reply deploy/seccompagent.yaml:72]: Cool. Can you please add a comment saying this path is needed for the freeze handler only?

File: pkg/handlers/default.go (hunk 3 above)
Comment (rata) [reply pkg/handlers/default.go]: Cool

File: pkg/handlers/default.go (hunk 9 above)
Comment (rata) [reply pkg/handlers/default.go:135]: Ohh, I see how the return is used. It is used for the response for the syscall. Looks fine then, thanks!

**Comment by rata** [comment]
@alban thanks! Most are solved, one small thing and more questions about if we really need a regex (it seems we don't and I'd like to avoid it if possible). Ping me when you answer, in case I miss the notification :)

**Comment by alban** [comment]
@rata Branch updated, this time without regex!
